File: query/src/main/java/com/quarkworks/android/realmtypesafequery/fields/RealmIntegerField.java
package com.quarkworks.android.realmtypesafequery.fields;

import android.support.annotation.NonNull;

import com.quarkworks.android.realmtypesafequery.RealmField;
import com.quarkworks.android.realmtypesafequery.RealmFieldType;
import com.quarkworks.android.realmtypesafequery.converter.RealmFieldConverter;

import io.realm.RealmModel;

public class RealmIntegerField<Model extends RealmModel, Value> extends RealmField<Model, Integer, Value> {

    @NonNull
    public static <Model extends RealmModel> RealmIntegerField<Model, Integer> create(@NonNull Class<Model> modelType, @NonNull String name) {
        return new RealmIntegerField<>(modelType, name, RealmIntegerFieldConverter.SHARED_INSTANCE);
    }

    @NonNull
    public static <Model extends RealmModel, Value> RealmIntegerField<Model, Value> create(@NonNull Class<Model> modelType, @NonNull String name, RealmFieldConverter<Integer, Value> converter) {
        return new RealmIntegerField<>(modelType, name, converter);
    }

    public RealmIntegerField(@NonNull Class<Model> modelType, @NonNull String name, @NonNull RealmFieldConverter<Integer, Value> attributeConverter) {
        super(modelType, name, attributeConverter, RealmFieldType.INTEGER);
    }

    public static class RealmIntegerFieldConverter implements RealmFieldConverter<Integer, Integer> {

        public static final RealmIntegerFieldConverter SHARED_INSTANCE = new RealmIntegerFieldConverter();

        @Override
        public Integer convertToValue(Integer attribute) {
            return attribute;
        }

        @Override
        public Integer convertToRealmValue(Integer value) {
            return value;
        }
    }
}

File: query/src/main/java/com/quarkworks/android/realmtypesafequery/fields/RealmIntegerField.kt
package com.quarkworks.android.realmtypesafequery.fields

import io.realm.RealmModel
import io.realm.RealmQuery

open class RealmIntegerField<Model : RealmModel>(override val modelClass: Class<Model>, override val name: String) :
        RealmField<Model>,
        RealmEquatableField<Model, Int>,
        RealmComparableField<Model, Int>,
        RealmSortableField<Model, Int>,
        RealmInableField<Model, Int>,
        RealmMinMaxField<Model, Int>,
        RealmAggregatableField<Model, Long> {

    override fun equalTo(query: RealmQuery<Model>, value: Int?) {
        if (value == null) {
            isNull(query)
            return
        }

        query.equalTo(name, value)
    }

    override fun notEqualTo(query: RealmQuery<Model>, value: Int?) {
        if (value == null) {
            isNotNull(query)
            return
        }

        query.notEqualTo(name, value)
    }

    override fun never(query: RealmQuery<Model>) {
        val zero = 0
        val one = 1
        query.beginGroup().equalTo(name, zero).equalTo(name, one).endGroup()
    }

    override fun greaterThan(query: RealmQuery<Model>, value: Int?) {
        if (value == null) {
            notEqualTo(query, null)
            return
        }

        query.greaterThan(name, value)
    }

    override fun greaterThanOrEqualTo(query: RealmQuery<Model>, value: Int?) {
        if (value == null) {
            equalTo(query, null)
            return
        }

        query.greaterThanOrEqualTo(name, value)
    }

    override fun lessThan(query: RealmQuery<Model>, value: Int?) {
        if (value == null) {
            notEqualTo(query, null)
            return
        }

        query.lessThan(name, value)
    }

    override fun lessThanOrEqualTo(query: RealmQuery<Model>, value: Int?) {
        if (value == null) {
            equalTo(query, null)
            return
        }

        query.lessThanOrEqualTo(name, value)
    }

    override fun between(query: RealmQuery<Model>, start: Int?, end: Int?) {
        if (start == null || end == null) {
            equalTo(query, null)
        }

        query.between(name, start!!, end!!)
    }

    override fun `in`(query: RealmQuery<Model>, values: Array<Int>) {
        query.`in`(name, values)
    }

    override fun min(query: RealmQuery<Model>): Int? = query.min(name)?.toInt()

    override fun max(query: RealmQuery<Model>): Int? = query.max(name)?.toInt()

    override fun sum(query: RealmQuery<Model>): Long = query.sum(name).toLong()
}
